### PR TITLE
istioctl upgrade: use yaml diff for upgrade iops

### LINKED ISSUE #28555: istioctl upgrade in-place has confusing output
```
$ istio-1.7.1/bin/istioctl install 
$ istio-1.8.1-alpha.0/bin/istioctl upgrade
```

The output below has a couple of issues:
1. it is testing changes in components like egressGateways and kiali which are not installed. This generates a tremendous amount of irrelevant output which obscures the useful information.
2. entries like
              '[?->0]': -> map[name:tcp-citadel-grpc-tls port:8060 targetPort:8060]
need to be explained
3. kiali and prometheus are deprecated and should be removed completely from the diff. 

```
Control Plane - ingressgateway pod - istio-ingressgateway-dbfcf54b8-z4sdm - version: 1.7.1
Control Plane - pilot pod - istiod-684dc4d6d6-dc5qn - version: 1.7.1
Control Plane - pilot pod - istiod-684dc4d6d6-p5kj7 - version: 1.7.1

Upgrade version check passed: 1.7.1 -> 1.8.0.

Upgrade check: Warning!!! The following IOPS will be changed as part of upgrade. Please double check they are correct:
spec:
  components:
    egressGateways:
      '[#0]':
        k8s:
          env:
            '[#0]':
              value: sni-dnat -> standard
          service:
            ports:
              '[#0]':
                protocol: -> TCP
              '[#1]':
                protocol: -> TCP
              '[#2]':
                protocol: -> TCP
    ingressGateways:
      '[#0]':
        k8s:
          env:
            '[#0]':
              value: sni-dnat -> standard
          resources:
            requests:
              cpu: 100m -> 200m
          service:
            ports:
              '[?->0]': -> map[name:tcp-citadel-grpc-tls port:8060 targetPort:8060]
              '[?->1]': -> map[name:tcp-dns port:5353]
              '[0->?]': map[name:status-port port:15021 targetPort:15021] ->
              '[1->?]': map[name:http2 port:80 targetPort:8080] ->
              '[2->?]': map[name:https port:443 targetPort:8443] ->
              '[3->?]': map[name:tls port:15443 targetPort:15443] ->
          serviceAnnotations: -> map[cloud.google.com/load-balancer-type:internal]
    pilot:
      k8s:
        hpaSpec: -> map[maxReplicas:10 minReplicas:2]
        overlays: -> [map[kind:Deployment name:istiod patches:[map[path:spec.template.spec.containers.[name:discovery].args.[30m] value:60m] map[path:spec.template.spec.containers.[name:discovery].ports.[containerPort:8080].containerPort value:1234] map[path:spec.template.spec.containers.[name:discovery].env.[name:REVISION]] map[path:spec.template.spec.containers.[name:discovery].securityContext]]] map[kind:Service name:istiod patches:[map[path:spec.ports.[name:https-dns].port value:11111]]]]
        resources: -> map[requests:map[cpu:1000m memory:4096Mi]]
  installPackagePath: /tmp/istio-install-packages/istio-1.7.1/manifests ->
  values:
    gateways:
      istio-ingressgateway:
        debug: info ->
        meshExpansionPorts: '[map[name:tcp-istiod port:15012 targetPort:15012] map[name:tcp-dns-tls port:853 targetPort:8853]] ->'
    global:
      controlPlaneSecurityEnabled: true ->
      enableHelmTest: false ->
      tracer:
        datadog:
          address: $(HOST_IP):8126 ->
        stackdriver:
          debug: false ->
          maxNumberOfAnnotations: 200 ->
          maxNumberOfAttributes: 200 ->
          maxNumberOfMessageEvents: 200 ->
      trustDomain: cluster.local ->
    kiali: map[contextPath:/kiali createDemoSecret:false dashboard:map[auth:map[strategy:login] grafanaInClusterURL:http://grafana:3000 grafanaURL:<nil> jaegerInClusterURL:http://tracing/jaeger jaegerURL:<nil> passphraseKey:passphrase secretName:kiali usernameKey:username viewOnlyMode:false] hub:quay.io/kiali nodeSelector:map[] podAntiAffinityLabelSelector:[] podAntiAffinityTermLabelSelector:[] prometheusNamespace:<nil> security:map[cert_file:/kiali-cert/cert-chain.pem enabled:false private_key_file:/kiali-cert/key.pem] service:map[annotations:map[]] tag:v1.22] ->
    ? map[accessMode:ReadWriteMany contextPath:/grafana dashboardProviders:map[dashboardproviders.yaml:map[apiVersion:1 providers:[map[disableDeletion:false folder:istio name:istio options:map[path:/var/lib/grafana/dashboards/istio] orgId:1 type:file]]]] datasources:map[datasources.yaml:map[apiVersion:1 datasources:<nil>]] env:map[] envSecrets:map[] image:map[repository:grafana/grafana tag:7.0.5] nodeSelector:map[] persist:false podAntiAffinityLabelSelector:[] podAntiAffinityTermLabelSelector:[] security:map[enabled:false passphraseKey:passphrase secretName:grafana usernameKey:username] service:map[annotations:map[] externalPort:3000 loadBalancerIP:<nil> loadBalancerSourceRanges:<nil> name:http type:ClusterIP] storageClassName
    : tolerations:[]] ->
    ? map[adapters:map[kubernetesenv:map[enabled:true] prometheus:map[enabled:true metricsExpiryDuration:10m] stackdriver:map[auth:map[apiKey
    : appCredentials:false serviceAccountPath:] enabled:false tracer:map[enabled:false sampleProbability:1]] stdio:map[enabled:false outputAsJson:false] useAdapterCRDs:false] policy:map[adapters:map[kubernetesenv:map[enabled:true] useAdapterCRDs:false] autoscaleEnabled:true image:mixer sessionAffinityEnabled:false] telemetry:map[autoscaleEnabled:true env:map[GOMAXPROCS:6] image:mixer loadshedding:map[latencyThreshold:100ms mode:enforce] nodeSelector:map[] podAntiAffinityLabelSelector:[] podAntiAffinityTermLabelSelector:[] replicaCount:1 sessionAffinityEnabled:false tolerations:[]]] ->
    ? map[jaeger:map[accessMode:ReadWriteMany hub:docker.io/jaegertracing memory:map[max_traces:50000] persist:false spanStorageType:badger storageClassName
    : tag:1.18] nodeSelector:map[] opencensus:map[exporters:map[stackdriver:map[enable_tracing:true]] hub:docker.io/omnition resources:map[limits:map[cpu:1 memory:2Gi] requests:map[cpu:200m memory:400Mi]] tag:0.1.9] podAntiAffinityLabelSelector:[] podAntiAffinityTermLabelSelector:[] provider:jaeger service:map[annotations:map[] externalPort:9411 name:http-query type:ClusterIP] zipkin:map[hub:docker.io/openzipkin javaOptsHeap:700 maxSpans:500000 node:map[cpus:2] probeStartupDelay:10 queryPort:9411 resources:map[limits:map[cpu:1000m memory:2048Mi] requests:map[cpu:150m memory:900Mi]] tag:2.20.0]] ->
    pilot:
      configNamespace: istio-config ->
      policy: map[enabled:false] ->
      traceSampling: 1 -> 0.1
    prometheus: map[contextPath:/prometheus hub:docker.io/prom nodeSelector:map[] podAntiAffinityLabelSelector:[] podAntiAffinityTermLabelSelector:[] provisionPrometheusCert:true retention:6h scrapeInterval:15s security:map[enabled:true] tag:v2.19.2 tolerations:[]] ->
    sidecarInjectorWebhook:
      injectLabel: istio-injection ->
    telemetry:
      v1: map[enabled:false] ->
```

### PR DESCRIPTION
**Pull request by shamsher31**
To fix https://github.com/istio/istio/issues/28555

Based on https://github.com/istio/istio/issues/28555#issuecomment-721655199

```
[istio-1.7.3] $ istioctl install 
[istio-1.8.0-alpha.2] $ istioctl upgrade
```
Before:

```
    ingressGateways:
      '[#0]':
        k8s:
          env:
            '[#0]':
              value: sni-dnat -> standard
          resources:
            requests:
              cpu: 100m -> 200m
          service:
            ports:
              '[?->0]': -> map[name:tcp-citadel-grpc-tls port:8060 targetPort:8060]
              '[?->1]': -> map[name:tcp-dns port:5353]
              '[0->?]': map[name:status-port port:15021 targetPort:15021] ->
              '[1->?]': map[name:http2 port:80 targetPort:8080] ->
              '[2->?]': map[name:https port:443 targetPort:8443] ->
              '[3->?]': map[name:tls port:15443 targetPort:15443] ->
          serviceAnnotations: -> map[cloud.google.com/load-balancer-type:internal]
```

After:

colored-output

### CODE REVIEW

File: operator/cmd/mesh/upgrade.go [hunk 1]
```diff
@@ -167,22 +165,6 @@ func upgrade(rootArgs *rootArgs, args *upgradeArgs, l clog.Logger) (err error) {
 	}
 	l.LogAndPrintf("Upgrade version check passed: %v -> %v.\n", currentVersion, targetVersion)
 
-	// Read the overridden IOP from args.inFilenames
-	overrideIOPYaml := ""
-	if args.inFilenames != nil {
```
Comment (shamsher31) [inline operator/cmd/mesh/upgrade.go:172]: I just realized this might be required if user use `-f` for `upgrade`?

**Comment by su225** [comment]
@shamsher31  - Just a minor suggestion. While showing diffs, does it make sense to color them like git does? It looks better than searing for small `+` or `-` sign at the start of the line.

**Comment by shamsher31** [comment]
@su225 I will definitely prefer the colored output but not sure if the diff lib accepts any parameter to decorate the output.
https://github.com/kylelemons/godebug/tree/master/diff

**Comment by su225** [comment]
@shamsher31 - It is possible. Output is a string and it is well-documented that it would be either `+`, `-` in the library. As a nice to have it is possible to split the lines and put it ourselves in case we could not find a good library. This is generic enough that we could use it for other purposes like profile diffs (whatever diff, basically). As coloring is nice to have, we could leave it by creating an issue if this is urgent.

**Comment by shamsher31** [comment]
@su225 Thanks. added your suggestion for colored output https://github.com/istio/istio/pull/28572#issue-515305641

**Comment by su225** [comment]
@shamsher31 : Looks nice. Just a nitpick - why additional newline?

**Comment by shamsher31** [comment]
@ostromart PTAL.

File: operator/cmd/mesh/upgrade.go (hunk 1 above)
Comment (ostromart) [inline operator/cmd/mesh/upgrade.go:172]: i don't think we can delete this. upgrade uses the user config to not bother diffing values that are changed by the user (and which therefore are not affected by change in defaults).

**Comment by ostromart** [comment]
using yamldiff is not going to work well - it was actually used before the compare package was written, but the drawback that it has is that it doesn't give the full tree path of changed elements. so you end up seeing the leaf changed without seeing where in the tree that happened. a better approach would be to fix the relatively minor issues with compare to get a nicer output.

**Review by ostromart — CHANGES_REQUESTED** [review]

**Comment by shamsher31** [comment]
> using yamldiff is not going to work well - it was actually used before the compare package was written, but the drawback that it has is that it doesn't give the full tree path of changed elements. so you end up seeing the leaf changed without seeing where in the tree that happened.

I agree that it displays full YAML https://github.com/istio/istio/issues/28555#issuecomment-721655199, but that is the whole point of this PR because it highlights which part has been removed (- in red color) and which has been newly added(+ in green color). so the user can take better decisions, instead of just showing not so readable output where it is hard to identify changes in YAML.

**Comment by ostromart** [comment]
if you can figure out how to display the full path for yamldiff then i'd be ok with this approach. but we can't have "floating" value deltas that show (say) cpu setting changed without giving a context of which component it changed for. 
i think it would be much more fruitful to just fix the relatively minor formatting issues in the compare package rather than trying to fix this problem in yamldiff.

**Pull request closed without merging**